### ISSUE TITLE
Support/conflict code not consistent with documentation

### ISSUE THREAD

**Issue by jar398** [OpenTreeOfLife/germinator#50]
I'm seeing results from the tree_of_life API that are not consistent with the documentation. The documentation says the value of the property should be a list, but the code returns a dictionary. I don't know which is incorrect. Could someone (@bredelings @josephwb @mtholder) please explain how this inconsistency should be fixed?

The documentation is here: https://github.com/OpenTreeOfLife/opentree/wiki/Conflict-and-synthesis-file-formats  Sample output from the implementation is below.

Also it looks like there is only the possibility of being supported by a single node in a give tree, not by multiple nodes.  If there will never be more than one supporting node then I don't see why. Also I don't have a conflict example and don't know what it should look like, and there can *definitely* be more than one conflicting node) - so would there be a list of source node ids, instead of a single one?

    curl -X POST https://test.opentreeoflife.org/v3/tree_of_life/node_info \
    -H "content-type:application/json" -d '{"node_id":"mrcaott3504ott396446"}'
    > 
    {
      "synth_data" : {
        "supported_by" : {
          "pg_2539_tree6294" : "node1094822",
          "pg_1944_tree3959" : "node741000"
        },
        "source_id_map" : {
          "pg_2820_tree6566" : {
            "git_sha" : "c6ce2f9067e9c74ca7b1f770623bde9b6de8bd1f",
            "tree_id" : "tree6566",
            "study_id" : "pg_2820"
          },
          "pg_2539_tree6294" : {
            "git_sha" : "c6ce2f9067e9c74ca7b1f770623bde9b6de8bd1f",
            "tree_id" : "tree6294",
            "study_id" : "pg_2539"
          },
          "pg_1944_tree3959" : {
            "git_sha" : "c6ce2f9067e9c74ca7b1f770623bde9b6de8bd1f",
            "tree_id" : "tree3959",
            "study_id" : "pg_1944"
          }
        },
        "terminal" : {
          "pg_2820_tree6566" : "ott729871"
        },
        "tip_descendants" : 2637,
        "synth_id" : "opentree4.1"
      },
      "taxon" : {
      },
      "node_id" : "mrcaott3504ott396446"
    }

**Comment by jar398** [comment]
The value for "terminal" looks wrong, too - isn't it supposed to be an id for a node in the source tree?

**Comment by josephwb** [comment]
Regarding `terminal: that is currently what comes through the annotations json. Agree that doesn't look right.

**Comment by josephwb** [comment]
The (outer) dictionary bit ("synth_data") comes from the multi-tree idea: that a node could potentially occur in multiple synthesis trees. If true, the function would report all such synth-tree-specific instances. I will take this out since 1) it is not desired behaviour (although I personally think it would be useful), and (much more importantly) 2) non-taxon nodes across synthesis trees can potentially have the same `mrcaottXXXottYYY`label.
I don't see any mention of "list" in the page you linked to, but can change it if that is desired. The structure above looks intuitive to mean, but I don't know what people want or what is commonly done.

**Comment by josephwb** [comment]
Oh, _each_ individual thing is a list. I see.

**Comment by bredelings** [comment]
The JSON that I'm making looks a fair bit different than this.
```
        "mrcaott3504ott396446": {
            "terminal": [
                [
                    "pg_2820_tree6566", 
                    "ott729871"
                ]
            ], 
            "supported_by": [
                [
                    "pg_1944_tree3959", 
                    "node741000"
                ], 
                [
                    "pg_2539_tree6294", 
                    "node1094822"
                ]
            ]
        }, 
```

**Comment by bredelings** [comment]
We also get bits that look like this:
``` json
        "ott716639": {
            "terminal": [
                [
                    "pg_2690_tree6243", 
                    "ott716639"
                ]
            ]
        }, 
```

**Comment by bredelings** [comment]
I'm not sure why the "mrcaott3504ott396446" would even _be_ a terminal at all.
For "ott716639", yeah, I think we should mention the node name w/in the study tree.

**Comment by bredelings** [comment]
Joseph, I know you suggested that we change the [studytree, node] arrays to studytree:node elements.  I guess you are changing the arrays to key:value pairs inside tm-lite.  Let's just preserve the original array form for now.  That's what's documented, let's go with that.

**Comment by josephwb** [comment]
Yeah, I will change things. The problem is that neo4j cannot store nested values, so I need to reformat things on ingestion, and then reformat them again on export.

**Comment by jar398** [comment]
"ott716639" doesn't make sense for terminal annotations. Terminal should be
just like all the others: you want to know which is the relevant (terminal)
node in the source tree.  "ott716639" doesn't identify a node in the source
tree.

Besides, it's redundant with information provided as the enclosing
dictionary key. If the source tree node id has been thrown away, just
provide the source id, not source id + repeated OTT id.

**Comment by jar398** [comment]
I don't like the phrase "for now". We are likely to be stuck with the
chosen format for a long time. So it's worth a little thought.

Concatenating the source tree id (which "for now" is an opaque token,
possibly with restricted character set) with the node id using '_' as a
separator is a very bad idea for reasons I've stated before.

Maybe it is time for a more orderly review of the syntactic options.
@josephwb you are of course free to change the code, but expect that it
might have to change again, since we are in the middle of review.

It's more important to me that the choices be thought through and defended,
than that they be to my taste.

**Comment by bredelings** [comment]
@jar398 Regarding `terminal`, I agree that 
* terminal should be like the others - you want to know which is the relevant node in the source tree
* adding ott716639 is redundant with the dictionary key.
* I will fix otc-annotate-synth to mention nodeXXXX instead of ottXXXX when identifying study tree nodes that are terminals.
So, I think we are in sufficient agreement.  Am I missing something?

**Comment by jar398** [comment]
Yes, sounds like we agree.

**Comment by mtholder** [comment]
wrt the number of possible "supported_by" nodes from a study: yes there is a maximum one (for the definition of supported_by that I've been lobbying for). 

That said, I would not object to making it a list of one object, if that would make the blob more similar for the conflict API.

**Comment by josephwb** [comment]
What about "partial_path_of" or "conflicts": couldn't they have more than one node from a study tree? Maybe this is what @jar398 meant?

**Comment by jar398** [comment]
Then why do we have this (in https://github.com/OpenTreeOfLife/clade-workshops/blob/master/synthesis_trees/belidae_collection/synth_2016-02-22_run3/annotations.json ):

       "mrcaott142444ott3407764": {
            "supported_by": [
                [
                    "ot_626_tree1", 
                    "node12"
                ], 
                [
                    "ot_626_tree1", 
                    "node11"
                ]
            ], 

I understand that only one synthetic tree node can be supported by a given source tree node, by definition of supported_by, but I believe it's possible that a synthetic tree node can be supported by multiple nodes in a single source tree.  (Congruence is many-to-many, supported_by is one-to-many.)

**Comment by bredelings** [comment]
Interesting.  Apparently if you look at the nexson, there is a knuckle:

node5-> node11 -> node12

-BenRI

**Comment by mtholder** [comment]
Intere

**Comment by jar398** [comment]
That approach would work, but only if you first pruned off non-taxonomy tips of the source tree.  Which you do, so it would be fine for purposes of this synth pipeline, and maybe it's OK to exploit that in the way the supported_by annotations are expressed (i.e. skip the singleton list enclosure) and in what the node_info API method returns. (But in a broader context (e.g. the conflict tool I wrote) it would have to be acknowledged that a node in tree 1 can be supported by multiple non-knuckle nodes in tree 2.)

**Comment by jar398** [comment]
There are really two things going on here, synthesis output and tm-lite (API) output.
I've updated the synthesis output documentation: https://github.com/OpenTreeOfLife/opentree/wiki/Conflict-and-synthesis-file-formats and will take the implementation as authoritative. So the propinquity part of this issue is finished. I've clarified that the wiki page is about synthesis output, *not* the web API. The web API might or might not use the same support/conflict information format as the synthesis output JSON. Since no one is taking a stand on what they think the API should do, and Joseph has requested a spec, I'm assigning this issue to me and I will come up with something, taking advice from synthesis output, the current tm-lite implementation, and Mark's August 2015 v3 API notes.
